Celeste.Mod.mm/Patches/Monocle/Tracker.cs

```diff
@@ -71,57 +73,116 @@ private static List<Type> GetSubclasses(Type type) {
+                }
+                // do the same for subclasses
+                foreach (Type subtype in subtypes) {
+                    if (!subtype.IsAbstract) {
+                        if (!TrackedComponentTypes.TryGetValue(subtype, out List<Type> value)) {
+                            value = new List<Type>();
+                            TrackedComponentTypes.Add(subtype, value);
                         }
-                    } else {
-                        // this is neither an entity nor a component. Help!
-                        throw new Exception("Type '" + type.Name + "' cannot be TrackedAs because it does not derive from Entity or Component");
+                        value.AddRange(TrackedComponentTypes.TryGetValue(trackedAsType, out List<Type> list) ? list : new List<Type>());
+                        TrackedComponentTypes[subtype] = value.Distinct().ToList();
                     }
                 }
             }
+            else {
+                // this is neither an entity nor a component. Help!
+                throw new Exception("Type '" + type.Name + "' cannot be Tracked because it does not derive from Entity or Component");
+            }
+        }
 
-            // don't hold references to all the types anymore
-            _temporaryAllTypes = null;
+        /// <summary>
+        /// Ensures the current scene's tracker contains all entities of all tracked Types.
+        /// Must be called if a type is added to the tracker manually and if the active scene changes.
+        /// If called back to back without a type added to the Tracker, it won't go through again, for performance.
+        /// <paramref name="force"/> will make ensure the Refresh happens, even if run back to back.
+        /// Due to only the active Scene's Tracker's refreshed state is changed, <paramref name="force"/> must be true if a different scene becomes active.
+        /// </summary>
+        public void Refresh(bool force = false) {
+            if (!Unrefreshed && !force) {
+                return;
+            }
+            Unrefreshed = false;
+            foreach (Type entityType in StoredEntityTypes) {
+                if (!Entities.ContainsKey(entityType)) {
+                    Entities.Add(entityType, new List<Entity>());
+                }
+            }
+            foreach (Type componentType in StoredComponentTypes) {
+                if (!Components.ContainsKey(componentType)) {
+                    Components.Add(componentType, new List<Component>());
+                }
+            }
+            foreach (Entity entity in Engine.Scene.Entities) {
+                foreach (Component component in entity.Components) {
+                    Type componentType = component.GetType();
+                    if (Components[componentType].Contains(component)
+                        || !TrackedComponentTypes.TryGetValue(componentType, out List<Type> componentTypes)) {
```

Comment:
One final perf thing, I'm pretty sure `Components[componentType]` is a List, so doing a Contains check on it is slower than the dictionary check after the `||` operator. I think flipping the order of these 2 conditions will be a bit better for perf. (Same applies to entity code below)